### PR TITLE
#162685677 User can bookmark articles

### PR DESCRIPTION
**Pull request by fahadmak**
#### What does this PR do?
PR enables users should be able to bookmark an article.
#### Description of Task to be completed?
A logged in/authenticated user should be able to bookmark articles. The user should also be able to get all his/her bookmarks and to get or delete a single bookmark.
#### How should this be manually tested?
Before testing run these commands in cmd

`git clone https://github.com/andela/Ah-backend-guardians.git`

`cd Ah-backend-guardians`

`git checkout ft-bookmark-article-162685677`

`virtualenv env`

`. env/bin/activate`

`pip install -r requirements.txt`

1.  Create a Postgres database and a `.env` file in your root directory and add the following configurations.
```
SECRET_KEY=andela_sjdndkjghdf78872

DB_NAME=your_database_name

DB_USER=database_user

DB_PASSWORD=database_password

DB_HOST=127.0.0.1

ALLOWED_HOSTS=.localhost, .herokuapp.com

EMAIL_PORT=587

EMAIL_HOST_USER = gauthorshaven@gmail.com

EMAIL_HOST_PASSWORD = guardiansauthors

WORD_LENGTH = 5

WPM = 200
```
2. Copy and paste this link into postman http://127.0.0.1:8000/api/articles/ to test the app using postman. First, create the article by Sending POST request using this data below in body section
```
{
	
	"title": "An article",
	"body":"This is an article that does so many things",
	"description":"This is an article"
}
```

3. Then Go over to the drop-down and select the GET request to return all article's where you will the copy the slug of the article you created.
**This should be expected response** 
```{
    "article": {
        "id": 24,
        "title": "An articles534w",
        "body": "This is an article that does so many things",
        "description": "This is an article",
        "author": {
            "email": "paulviv112@gmail.com",
            "username": "pault211"
        },
        "slug": "an-articles",
        "published": true,
        "created_at": "2019-01-30T13:55:33.427626Z",
        "updated_at": "2019-01-30T13:55:33.427654Z",
        "images": "",
        "read_time": 0,
        "tags": [],
        "average_rating": 0,
        "social_links": {
            "facebook": "https://www.facebook.com/sharer/sharer.php?u=http%3A//127.0.0.1%3A8000/api/articles/an-articles534w-11/",
            "twitter": "https://twitter.com/home?status=http%3A//127.0.0.1%3A8000/api/articles/an-articles534w-11/",
            "email": "mailto:?&subject=An%20articles534w%20from%20Author%27s%20Haven&body=Click%20Link%20To%20View%20The%20Article%20http%3A//127.0.0.1%3A8000/api/articles/an-articles534w-11/"
        },
        "likes_count": 0,
        "dislikes_count": 0,
        "favourited": false,
        "favouriteCount": 0
    }
}
```
4. Then Copy and paste this link into postman http://127.0.0.1:8000/api/article/an-articles/bookmark/  Send POST request. The bookmark is then created.
**This should be expected response** 
```{
    "bookmarks": {
        "id": 29,
        "title": "An articles534w",
        "description": "This is an article",
        "read_time": 0,
        "author": "pault211",
        "created_at": "2019-01-30T14:15:53.415710Z",
        "reader": 1
    }
}
```
#### What are the relevant pivotal tracker stories?
[#162685677](https://www.pivotaltracker.com/story/show/162685677)

### CODE REVIEW

**Review by VivianDoreen — COMMENTED** [review]
Kindly conform to pep8 style guide

File: authors/apps/articles/serializers.py [hunk 1]
```diff
@@ -101,3 +101,33 @@ class Meta:
...
+    read_time = serializers.SerializerMethodField()
+    author = serializers.SerializerMethodField()
+
+    class Meta:
+        model = Bookmark
+        fields = ("id", "title", "description",
+                  "read_time", "author", "created_at", "reader")
+
+    def get_author(self, obj):
+        return obj.article.author
```
Comment (VivianDoreen) [inline authors/apps/articles/serializers.py]: Remove one of the get_author functions, it's a repetition

File: authors/apps/articles/renderers.py [hunk 2]
```diff
@@ -39,3 +39,31 @@ def render(self, data, media_type=None, renderer_context=None):
...
+        if 'errors' in data or 'detail' in data or 'error' in data:
+
+            return super().render(data)
+
+        return json.dumps({
+            'bookmark': data
+        })
+
+
+class BookmarkJSONRenderer(JSONRenderer):
```
Comment (VivianDoreen) [inline authors/apps/articles/renderers.py]: Remove one of the functions above

File: authors/apps/articles/views.py [hunk 3]
```diff
@@ -365,3 +389,74 @@ def get(self, request, *args, **kwargs):
...
+            bookmark = self.get_queryset().filter(article=article).first()
+            if bookmark:
+                raise ValidationError("This article is already bookmarked")
+            bookmark = Bookmark.objects.create(
+                article=article, reader=request.user)
+            serializer = self.serializer_class(bookmark)
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+        except Article.DoesNotExist:
+            raise NotFound(
+                {'error': 'Article with does not exist'})
```
Comment (VivianDoreen) [inline authors/apps/articles/views.py]: Kindly correct the grammar

File: authors/apps/articles/views.py [hunk 4]
```diff
@@ -365,3 +389,74 @@ def get(self, request, *args, **kwargs):
...
+    renderer_classes = (BookmarkJSONRenderer,)
+
+    def get_queryset(self):
+        return self.queryset.filter(reader=self.request.user)
+
+    def post(self, request, *args, **kwargs):
+        slug = self.kwargs.get('slug')
+        try:
+            article = Article.objects.get(slug=slug)
+            bookmark = self.get_queryset().filter(article=article).first()
```
Comment (VivianDoreen) [inline authors/apps/articles/views.py]: Look into using the get_object() method or get_queryset() method

**Comment by Opio-Emmanuel-Omona** [comment]
image

In your PR I have tried POST and it doesn't seem to be working.
image

**Review by Opio-Emmanuel-Omona — APPROVED** [review]
LGTM

**Review by VivianDoreen — APPROVED** [review]
LGTM

**Pull request merged**